lib/synapse/service_watcher/multi/resolver/s3_toggle.rb

```diff
@@ -0,0 +1,229 @@
+        return @watcher_setting
+      }
+    end
+
+    # pick and set a new watcher, but only if the hash has changed
+    def set_watcher(watcher_weights)
+      if watcher_weights.hash == @last_watcher_weights_hash
+        log.info "synapse: s3 toggle resolver: watcher weights hash does not change; ignoring update"
+        statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:skip'])
+
+        return get_watcher
+      end
+
+      picked_watcher = pick_watcher(watcher_weights)
+      if picked_watcher.nil?
+        return get_watcher
+      end
+
+      @last_watcher_weights_hash = watcher_weights.hash
+      @watcher_mu.synchronize {
+        @watcher_setting = picked_watcher
+      }
+
+      return picked_watcher
+    end
+
+    # randomly pick a watcher based on the provided weights.
+    # if it returns nil, it means it could not pick.
+    def pick_watcher(watcher_weights)
+      if watcher_weights.nil?
+        log.warn "synapse: s3 toggle resolver: received nil watcher weights, not picking"
+        statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:fail', 'reason:nil_weights'])
+
+        return nil
+      end
+
+      # Filter out any watchers that do not exist
+      watcher_weights = watcher_weights.select do |watcher, _|
+        exists = @watchers.has_key?(watcher)
+        unless exists
+          log.warn "synapse: s3 toggle resolver: read invalid watcher name #{watcher}"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.unknown_watchers')
+        end
+
+        exists
+      end
+
+      watcher_name = if watcher_weights.length > 0
+        # Pick a watcher randomly by weight.
+        total_weight = watcher_weights.values.inject(:+)
+        pick = rand(total_weight)
+        chosen_watcher = nil
+
+        watcher_weights.each do |key, value|
+          if pick < value
+            chosen_watcher = key
+            break
+          else
+            pick -= value
+          end
+        end
+
+        if chosen_watcher.nil?
+          log.warn "synapse: s3 toggle resolver: failed to pick a watcher"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:fail', 'reason:no_choice'])
+
+          nil
+        else
+          log.info "synapse: s3 toggle resolver: chose watcher #{chosen_watcher}"
+          statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:success', "watcher:#{chosen_watcher}"])
+          chosen_watcher
+        end
+      else
+        log.warn "synapse: s3 toggle resolver: no watchers read, failed to pick a watcher"
+        statsd_increment('synapse.watcher.multi.resolver.s3_toggle.switch', ['result:fail', 'reason:watchers_missing'])
+
+        nil
+      end
+
+      return watcher_name
+    end
+
+    def read_s3_file
+      s3 = AWS::S3::Client.new
+
+      data =
+        begin
+          resp = s3.get_object(bucket: @s3_bucket, key: @s3_path)
```

Comment:
just confirm this is the behavior we want: if `get_object` failed, we will skip this poll and wait for next one rather than sleep and retry, right?

Reply:
Yes, that's correct. That being said, the S3 client itself will retry certain types of errors.